File: modules/herdr-ssh/android/src/main/java/expo/modules/herdrssh/SshConnection.kt
package expo.modules.herdrssh

import android.util.Base64
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import net.schmizz.sshj.SSHClient
import net.schmizz.sshj.common.Buffer
import net.schmizz.sshj.common.IOUtils
import net.schmizz.sshj.transport.verification.HostKeyVerifier
import net.schmizz.sshj.userauth.UserAuthException
import net.schmizz.sshj.userauth.password.PasswordUtils
import java.io.IOException
import java.security.MessageDigest
import java.security.PublicKey

/** A failure the caller is expected to handle. `code` matches `SshFailureCode` in TypeScript. */
class SshFailure(
  val code: String,
  override val message: String,
  /** On `host_key_changed`: the fingerprint the host presented instead of the pin. */
  val presentedFingerprint: String? = null,
) : Exception(message)

/**
 * One long-lived SSH connection to a herdr host.
 *
 * Ported from the Compose app's `core.net.SshTransport`. The connection is
 * opened lazily and reused; a cached client is liveness-checked before use, and
 * a command that fails at the connection level drops the client. The next call
 * reconnects, but a possibly applied command is never automatically replayed.
 */
class SshConnection(private val config: SshConfigRecord) {

  private val mutex = Mutex()
  @Volatile private var client: SSHClient? = null
  /** The client being dialled, so a deadline can close it without the mutex. */
  @Volatile private var dialing: SSHClient? = null
  @Volatile private var hostKeyMismatch = false
  @Volatile private var presentedFingerprint: String? = null
  @Volatile var acceptedFingerprint: String? = null
    private set

  suspend fun connected(): SSHClient = mutex.withLock {
    client?.takeIf { it.isConnected && it.isAuthenticated }?.let { return it }
    runCatching { client?.disconnect() }
    client = null

    withContext(Dispatchers.IO) {
      hostKeyMismatch = false
      val fresh = SSHClient()
      fresh.addHostKeyVerifier(hostKeyVerifier())
      // Bounds the TCP handshake itself, which `connect` from JavaScript calls
      // with no deadline of its own. Not `fresh.timeout`: that is SO_TIMEOUT on
      // the socket, and sshj's reader would give up on a quiet `tail -f`.
      fresh.connectTimeout = CONNECT_TIMEOUT_MS
      dialing = fresh
      try {
        fresh.connect(config.host, config.port)
        // Survive NAT/router idle timeouts on a long-lived connection.
        fresh.connection.keepAlive.keepAliveInterval = 20
        authenticate(fresh)
        client = fresh
        fresh
      } catch (error: Exception) {
        runCatching { fresh.disconnect() }
        if (hostKeyMismatch) {
          throw SshFailure(
            "host_key_changed",
            "This host's SSH key has changed since you saved it.",
            presentedFingerprint,
          )
        }
        if (error is SshFailure) throw error
        throw SshFailure("connect_failed", error.message ?: "Couldn't reach the host.")
      } finally {
        dialing = null
      }
    }
  }

  /**
   * Trust-on-first-use: record the host key's SHA-256 fingerprint on first
   * contact; refuse any later connection presenting a different key.
   */
  private fun hostKeyVerifier(): HostKeyVerifier = object : HostKeyVerifier {
    override fun verify(hostname: String, port: Int, key: PublicKey): Boolean {
      // The SSH WIRE encoding of the key, not `key.encoded` — that is X.509
      // SPKI DER, a different blob, so it hashed to a fingerprint no other SSH
      // tool would ever print. OpenSSH's form is unpadded base64 of SHA-256
      // over the wire blob, which is what `ssh-keygen -lf` shows and what iOS
      // reports.
      val wire = Buffer.PlainBuffer().putPublicKey(key).compactData
      val digest = MessageDigest.getInstance("SHA-256").digest(wire)
      val fingerprint = Base64.encodeToString(digest, Base64.NO_WRAP).trimEnd('=')
      val pin = config.hostKeyFingerprint?.takeIf { it.isNotEmpty() } ?: acceptedFingerprint
      // Recorded only on acceptance. Assigning before the comparison published
      // a REJECTED key as the one we accepted, which is exactly the key a
      // "trust the new key" flow must never be handed for free.
      if (pin == null || pin == fingerprint) {
        acceptedFingerprint = fingerprint
        return true   // no pin: first contact, trust and record
      }
      hostKeyMismatch = true
      presentedFingerprint = fingerprint
      return false
    }

    override fun findExistingAlgorithms(hostname: String, port: Int): List<String> = emptyList()
  }

  private fun authenticate(client: SSHClient) {
    try {
      if (config.authKind == "password") {
        client.authPassword(config.username, config.password ?: "")
      } else {
        val passphrase = config.passphrase
          ?.takeIf { it.isNotEmpty() }
          ?.let { PasswordUtils.createOneOff(it.toCharArray()) }
        val keys = client.loadKeys(config.privateKey ?: "", null, passphrase)
        client.authPublickey(config.username, keys)
      }
    } catch (error: UserAuthException) {
      throw SshFailure(
        "auth_failed",
        "The server rejected these credentials. Check the username and the key or password.",
      )
    } catch (error: IOException) {
      throw SshFailure(
        "bad_key",
        "That private key couldn't be read. Paste the whole OpenSSH key (ed25519 or RSA), including the BEGIN/END lines, and check the passphrase.",
      )
    }
  }

  /**
   * Connect within a deadline. `connectTimeout` bounds only opening the socket;
   * a host that accepts TCP and then says nothing (not an SSH server, or a
   * wedged one) held the handshake, and Test connection with it, until the host
   * closed the socket, minutes later (#4 acceptance).
   */
  suspend fun connectWithin(timeoutMs: Int): SSHClient = withDeadline(timeoutMs) { connected() }

  suspend fun close() = mutex.withLock {
    withContext(Dispatchers.IO) { runCatching { client?.disconnect() } }
    client = null
  }

  private suspend fun resetClient() = mutex.withLock {
    withContext(Dispatchers.IO) { runCatching { client?.disconnect() } }
    client = null
  }

  /**
   * Drop the connection WITHOUT the mutex, for when something is stuck on it.
   *
   * `connected()` holds the mutex for as long as a dial takes, so a deadline
   * that waited for it would wait on the very thing that timed out. Closing
   * the socket is also what makes a read parked on it throw, which is the only
   * way to stop one: sshj's blocking reads ignore coroutine cancellation.
   * Also the module's teardown, which cannot suspend.
   */
  fun abandon() {
    val stuck = listOfNotNull(client, dialing)
    client = null
    dialing = null
    // Off the caller's thread: disconnect() writes a goodbye first, and a
    // write on a half-open socket can block too.
    if (stuck.isNotEmpty()) ioScope.launch { stuck.forEach { runCatching { it.disconnect() } } }
  }

  data class CommandOutput(val stdout: String, val stderr: String, val exitCode: Int)

  /**
   * Run once. A non-zero exit is a result. A lost reply is ambiguous, not
   * permission to execute a possibly applied command a second time.
   */
  suspend fun exec(command: String, timeoutMs: Int): CommandOutput =
    try {
      withDeadline(timeoutMs) { execOnce(command) }
    } catch (failure: SshFailure) {
      throw failure   // auth / host key / bad key / timeout: retrying changes nothing
    } catch (error: IOException) {
      resetClient()
      throw SshFailure("transport_failed", error.message ?: "The connection dropped.")
    }

  /**
   * Race an operation against its deadline.
   *
   * `withTimeout` alone was no deadline at all (#110): cancellation is
   * cooperative, sshj's reads block, and `withTimeout` waits for its block to
   * finish before throwing, so a read on a half-open socket held the command
   * forever. The operation now runs detached and only the WAIT is timed; on
   * expiry the connection is abandoned, which closes the socket under the
   * parked read so it fails and ends on its own. The next command dials fresh
   * rather than queueing behind a channel that will never answer.
   */
  private suspend fun <T> withDeadline(timeoutMs: Int, operation: suspend () -> T): T {
    if (timeoutMs <= 0) return operation()
    val work = ioScope.async { operation() }
    val outcome: Result<T>? = withTimeoutOrNull(timeoutMs.toLong()) {
      try {
        Result.success(work.await())
      } catch (cancelled: CancellationException) {
        throw cancelled
      } catch (error: Throwable) {
        Result.failure(error)
      }
    }
    if (outcome == null) {
      abandon()
      work.cancel()
      throw SshFailure(
        "timeout",
        "The host didn't answer in time. Check that it's awake and on the tailnet.",
      )
    }
    return outcome.getOrThrow()
  }

  private suspend fun execOnce(command: String): CommandOutput = withContext(Dispatchers.IO) {
    val client = connected()
    val session = client.startSession()
    try {
      val cmd = session.exec(command)
      // Both streams at once. Reading all of stdout before any of stderr
      // deadlocked a command that wrote enough to stderr to fill the channel's
      // window: it waited for stderr to drain, we waited for stdout to end.
      coroutineScope {
        val stderr = async(Dispatchers.IO) { IOUtils.readFully(cmd.errorStream).toString(Charsets.UTF_8.name()) }
        val stdout = IOUtils.readFully(cmd.inputStream).toString(Charsets.UTF_8.name())
        cmd.join()
        CommandOutput(stdout, stderr.await(), cmd.exitStatus ?: 0)
      }
    } finally {
      runCatching { session.close() }
    }
  }

  /**
   * Run a long-lived command and hand each stdout LINE to [onLine]. Returns a
   * handle whose `stop()` unblocks the parked reader and closes the channel.
   */
  suspend fun startStream(
    command: String,
    startTimeoutMs: Int,
    onLine: (String) -> Unit,
    onEnd: (Int) -> Unit,
    onError: (String, String) -> Unit,
  ): StreamHandle {
    // Only STARTING is bounded. What follows is a `tail -f` and is meant to
    // outlive any deadline; a stream that dies quietly is the tail watchdog's
    // job, not this one's.
    // Opening the channel is bounded too: it is a round-trip on a connection
    // that may have died since the last command.
    val (session, cmd) = withDeadline(startTimeoutMs) {
      val client = try {
        connected()
      } catch (failure: SshFailure) {
        throw failure
      } catch (error: IOException) {
        resetClient()
        connected()
      }
      val session = client.startSession()
      try {
        session to session.exec(command)
      } catch (error: Exception) {
        // The channel was open; without this it stayed open for the life of
        // the connection (#110).
        runCatching { session.close() }
        throw error
      }
    }

    return withContext(Dispatchers.IO) {
      val reader = cmd.inputStream.bufferedReader()
      val handle = StreamHandle(session)
      Thread {
        try {
          while (!handle.stopped) {
            val line = reader.readLine() ?: break
            onLine(line)
          }
          if (!handle.stopped) onEnd(cmd.exitStatus ?: 0)
        } catch (error: Exception) {
          if (!handle.stopped) {
            onError("transport_failed", error.message ?: "The stream ended unexpectedly.")
          }
        } finally {
          runCatching { session.close() }
        }
      }.apply { isDaemon = true }.start()
      handle
    }
  }

  private companion object {
    const val CONNECT_TIMEOUT_MS = 15_000
    /** Detached work for deadlines and teardown; never waited on by structure. */
    val ioScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  }

  class StreamHandle(private val session: net.schmizz.sshj.connection.channel.direct.Session) {
    @Volatile var stopped = false
      private set

    /**
     * Closing the session is what unblocks a `readLine()` parked on the channel.
     *
     * Off the caller's thread: sshj's close waits for the host to close its side,
     * and the caller is Expo's single queue for every async call of this module,
     * so a close that took 30 s held up every SSH call behind it (#4). The host
     * now ends the command when the channel closes (`untilChannelCloses`), which
     * keeps that wait short, but nothing should wait on it here.
     */
    fun stop() {
      stopped = true
      Thread { runCatching { session.close() } }.apply { isDaemon = true }.start()
    }
  }
}

File: modules/herdr-ssh/ios/SshConnection.swift
import Foundation
import Citadel
import Crypto
import NIOCore
import NIOSSH
import Network

/// A connection failure the caller is expected to handle rather than crash on.
/// Mirrors `SshFailureCode` in the TypeScript types — the string is the wire
/// contract between the two.
struct SshFailure: Error {
  let code: String
  let message: String

  static func connect(_ error: Error) -> SshFailure {
    // Citadel surfaces auth rejection as an authentication error and everything
    // else (DNS, refused, timeout) as a channel/IO error. The distinction
    // matters to the user: "wrong key" and "host unreachable" need different
    // fixes.
    let text = String(describing: error)
    if text.contains("authenticationFailed") || text.contains("allAuthenticationOptionsFailed") {
      return SshFailure(
        code: "auth_failed",
        message: "The server rejected these credentials. Check the username and the key or password."
      )
    }
    return SshFailure(code: "connect_failed", message: friendly(error))
  }

  static func friendly(_ error: Error) -> String {
    if let failure = error as? SshFailure { return failure.message }
    return (error as NSError).localizedDescription
  }
}

/// One long-lived SSH connection to a herdr host.
///
/// Ported from the SwiftUI app's `HerdrNet.SSHTransport`, which earned every one
/// of these behaviours in the field: the connection is opened lazily and reused
/// across commands, a cached client is liveness-checked before use, a command
/// that fails at the connection level drops the client without replaying a
/// possibly applied command, and the network path is
/// watched so a route change (wifi↔cellular, Tailscale up/down) invalidates the
/// socket instead of stalling on a dead one.
///
/// There is deliberately no SSH-level keepalive here, unlike the Kotlin side
/// which sets `keepAliveInterval = 20`: Citadel exposes no equivalent on
/// `SSHClientSettings`, so there is nothing to set. The path monitor above
/// covers a route change, and `isConnected` covers a client that noticed its own
/// death — but neither catches a channel silently dropped by a NAT idle timeout
/// on an unchanged route. Two things cover that, at two different layers:
///
/// `exec` carries a deadline (see `withDeadline`). Without one, a half-open
/// socket left the command pending forever, and because the app's poll loops
/// re-arm inside a `finally`, forever meant the loop stopped and the screen
/// froze holding stale data with no error shown.
///
/// A live `tail -f` cannot have an overall deadline — being long-lived is the
/// point — so it is covered instead by the tail watchdog in `useThread`, which
/// restarts a stream that has gone quiet while its agent is working. Only the
/// bounded part, getting the command started, has a deadline here.
actor SshConnection {
  private let config: SshConfigRecord
  private var client: SSHClient?
  /// The host key fingerprint this connection accepted, reported back to JS so
  /// first contact can be persisted as the pin.
  private let hostKeyPin: HostKeyPin
  var acceptedFingerprint: String? { hostKeyPin.fingerprint }

  private let pathMonitor = NWPathMonitor()
  private var pathMonitorStarted = false
  private var pathSignature: String?

  init(config: SshConfigRecord) {
    self.config = config
    self.hostKeyPin = HostKeyPin(config.hostKeyFingerprint)
  }

  // MARK: - Lifecycle

  func close() async {
    if let client { try? await client.close() }
    client = nil
    if pathMonitorStarted {
      pathMonitor.cancel()
      pathMonitorStarted = false
    }
  }

  private func resetClient() async {
    if let client { try? await client.close() }
    client = nil
  }

  /// Watch the network path: when the interface set changes, drop the cached
  /// client so the next command dials fresh on the new route rather than
  /// stalling on a now-dead socket.
  private func startPathMonitorIfNeeded() {
    guard !pathMonitorStarted else { return }
    pathMonitorStarted = true
    pathMonitor.pathUpdateHandler = { [weak self] path in
      let signature = "\(path.status)|" + path.availableInterfaces
        .map { "\($0.type)" }.sorted().joined(separator: ",")
      Task { await self?.pathChanged(to: signature) }
    }
    pathMonitor.start(queue: DispatchQueue(label: "dev.herdr.ssh.path"))
  }

  private func pathChanged(to signature: String) async {
    let previous = pathSignature
    pathSignature = signature
    if let previous, previous != signature {
      await resetClient()
    }
  }

  // MARK: - Connecting

  @discardableResult
  func connected() async throws -> SSHClient {
    startPathMonitorIfNeeded()
    // A dead cached client (dropped TCP) must not be reused.
    if let client {
      if client.isConnected { return client }
      await resetClient()
    }

    // Parse the key up front: the settings closure below cannot throw.
    // SSHAuthenticationMethod isn't Sendable but is effectively immutable once
    // built, so the capture is safe.
    nonisolated(unsafe) let auth = try Self.authMethod(for: config)
    let observer = HostKeyObserver(pin: hostKeyPin)

    let settings = SSHClientSettings(
      host: config.host,
      port: config.port,
      authenticationMethod: { auth },
      hostKeyValidator: .custom(TOFUHostKeyDelegate(observer: observer))
    )

    do {
      let newClient = try await SSHClient.connect(to: settings)
      client = newClient
      return newClient
    } catch {
      if observer.mismatched {
        throw SshFailure(
          code: "host_key_changed",
          message: "The server's SSH key DIFFERS from the saved one (possible MITM, or the server was reinstalled). If you trust it, edit and save the server to reset the pin."
        )
      }
      throw SshFailure.connect(error)
    }
  }

  private static func authMethod(for config: SshConfigRecord) throws -> SSHAuthenticationMethod {
    if config.authKind == "password" {
      return .passwordBased(username: config.username, password: config.password ?? "")
    }
    let pem = config.privateKey ?? ""
    let decryptionKey = config.passphrase.flatMap { $0.isEmpty ? nil : $0.data(using: .utf8) }
    // ed25519 is the modern default; fall back to RSA for older keys.
    if let key = try? Curve25519.Signing.PrivateKey(sshEd25519: pem, decryptionKey: decryptionKey) {
      return .ed25519(username: config.username, privateKey: key)
    }
    do {
      let rsa = try Insecure.RSA.PrivateKey(sshRsa: pem, decryptionKey: decryptionKey)
      return .rsa(username: config.username, privateKey: rsa)
    } catch {
      throw SshFailure(
        code: "bad_key",
        message: "That private key couldn't be read. Paste the whole OpenSSH key (ed25519 or RSA), including the BEGIN/END lines, and check the passphrase."
      )
    }
  }

  // MARK: - Host key pinning (TOFU)

  /// Written once on the SSH event loop during the handshake, read after
  /// `connect` returns or throws.
  private final class HostKeyObserver: @unchecked Sendable {
    let pin: HostKeyPin
    var mismatched = false

    init(pin: HostKeyPin) {
      self.pin = pin
    }
  }

  /// Trust-on-first-use: record the host key's SHA-256 fingerprint on first
  /// contact; refuse any later connection presenting a different key.
  private final class TOFUHostKeyDelegate: NIOSSHClientServerAuthenticationDelegate, @unchecked Sendable {
    private let observer: HostKeyObserver

    init(observer: HostKeyObserver) {
      self.observer = observer
    }

    func validateHostKey(hostKey: NIOSSHPublicKey, validationCompletePromise: EventLoopPromise<Void>) {
      var buffer = ByteBuffer()
      _ = hostKey.write(to: &buffer)
      let digest = SHA256.hash(data: Data(buffer.readableBytesView))
      // OpenSSH's format: unpadded base64 of SHA-256 over the wire encoding of
      // the key. Dropping the "=" is what makes this string equal to what
      // `ssh-keygen -lf` prints, so the two can be compared by eye.
      let fingerprint = Data(digest).base64EncodedString().replacingOccurrences(of: "=", with: "")
      if observer.pin.accept(fingerprint) {
        validationCompletePromise.succeed(())
      } else {
        observer.mismatched = true
        validationCompletePromise.fail(
          SshFailure(code: "host_key_changed", message: "host key fingerprint mismatch")
        )
      }
    }
  }

  // MARK: - Commands

  struct CommandOutput {
    let stdout: String
    let stderr: String
    let exitCode: Int
  }

  /// Run a command to completion. A non-zero exit is a RESULT, not an error:
  /// the caller decides what exit 127 means. Never replay a failed command:
  /// the host may have applied it before its reply was lost.
  func exec(_ command: String, timeoutMs: Int) async throws -> CommandOutput {
    do {
      return try await withDeadline(timeoutMs) { try await self.execOnce(command) }
    } catch let failure as SshFailure {
      if failure.code == "transport_failed" { await resetClient() }
      throw failure
    } catch {
      await resetClient()
      throw SshFailure(code: "transport_failed", message: SshFailure.friendly(error))
    }
  }

  /// Race an operation against its deadline.
  ///
  /// The losing task is cancelled, but a Citadel channel blocked on a half-open
  /// socket will not necessarily notice — so the client is dropped as well. The
  /// next command dials fresh rather than queueing behind a channel that is
  /// never going to answer.
  private func withDeadline<T: Sendable>(
    _ timeoutMs: Int,
    _ operation: @escaping @Sendable () async throws -> T
  ) async throws -> T {
    guard timeoutMs > 0 else { return try await operation() }
    do {
      return try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await operation() }
        group.addTask {
          try await Task.sleep(nanoseconds: UInt64(timeoutMs) * 1_000_000)
          throw SshFailure(
            code: "timeout",
            message: "The host didn't answer in time. Check that it's awake and on the tailnet."
          )
        }
        guard let first = try await group.next() else {
          throw SshFailure(code: "transport_failed", message: "The command produced no result.")
        }
        group.cancelAll()
        return first
      }
    } catch let failure as SshFailure where failure.code == "timeout" {
      await resetClient()
      throw failure
    }
  }

  private func execOnce(_ command: String) async throws -> CommandOutput {
    let client = try await connected()
    var stdout = ByteBuffer()
    var stderr = ByteBuffer()
    do {
      let stream = try await client.executeCommandStream(command)
      for try await chunk in stream {
        switch chunk {
        case .stdout(let buffer): stdout.writeImmutableBuffer(buffer)
        case .stderr(let buffer): stderr.writeImmutableBuffer(buffer)
        }
      }
      return CommandOutput(stdout: string(stdout), stderr: string(stderr), exitCode: 0)
    } catch let failed as SSHClient.CommandFailed {
      // The command ran and exited non-zero. That is an answer, and the output
      // it produced before exiting is part of it.
      return CommandOutput(stdout: string(stdout), stderr: string(stderr), exitCode: failed.exitCode)
    } catch let failure as SshFailure {
      throw failure
    } catch {
      throw SshFailure(code: "transport_failed", message: SshFailure.friendly(error))
    }
  }

  /// Run a long-lived command and hand each stdout LINE to `onLine`, then call
  /// `onEnd` (exit code) or `onError`. Returns once the command has started, so
  /// the caller can report a start failure synchronously.
  func startStream(
    _ command: String,
    startTimeoutMs: Int,
    onLine: @escaping @Sendable (String) -> Void,
    onEnd: @escaping @Sendable (Int) -> Void,
    onError: @escaping @Sendable (String, String) -> Void
  ) async throws -> Task<Void, Never> {
    let ready = AsyncThrowingStream<Void, Error>.makeStream()
    let task = Task {
      var buffer = Data()
      do {
        let client = try await connected()
        // Citadel's plain executeCommandStream has no cancellation cleanup.
        // withExec owns the channel and closes it when this task is cancelled.
        var started = false
        var readError: Error?
        do {
          try await client.withExec(command) { stream, _ in
            started = true
            do {
              try Task.checkCancellation()
              ready.continuation.finish()
              for try await chunk in stream {
                try Task.checkCancellation()
                guard case .stdout(let data) = chunk else { continue }
                buffer.append(contentsOf: data.readableBytesView)
                while let newline = buffer.firstIndex(of: 0x0A) {
                  let line = buffer[buffer.startIndex..<newline]
                  onLine(String(decoding: line, as: UTF8.self))
                  buffer.removeSubrange(buffer.startIndex...newline)
                }
              }
              try Task.checkCancellation()
            } catch { readError = error }
          }
        } catch ChannelError.alreadyClosed where started {
          // The host may close first. Citadel 0.12.1 closes again in withExec;
          // do not replace the read's real exit/cancellation with alreadyClosed.
        }
        if let readError { throw readError }
        if !buffer.isEmpty {
          onLine(String(decoding: buffer, as: UTF8.self))
        }
        onEnd(0)
      } catch let failed as SSHClient.CommandFailed {
        ready.continuation.finish(throwing: failed)
        if !buffer.isEmpty { onLine(String(decoding: buffer, as: UTF8.self)) }
        onEnd(failed.exitCode)
      } catch is CancellationError {
        ready.continuation.finish(throwing: CancellationError())
      } catch {
        ready.continuation.finish(throwing: error)
        onError("transport_failed", SshFailure.friendly(error))
      }
    }
    do {
      try await withDeadline(startTimeoutMs) {
        for try await _ in ready.stream { }
      }
      return task
    } catch {
      task.cancel()
      throw error
    }
  }

  private func string(_ buffer: ByteBuffer) -> String {
    String(decoding: buffer.readableBytesView, as: UTF8.self)
  }
}
